src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -355,19 +623,612 @@ private CategoryDataset buildDataSet(StaplerRequest req) {
+        boolean generateToolTip = true;
+        int count = 0;
+        for (CaseResult caseResult : tests) {
+            String caseName = caseResult.getFullName();
+            if (!allPackages && !caseName.startsWith(projectLevel))
+                continue;
+            count++;
+            caseName = caseResult.getName();
+            if (!generateToolTip)
+                continue;
+            else if (toolTipString.length() + caseName.length() > maxToolTipLength) {
+                generateToolTip = false;
+                toolTipString += ",...";
+                continue;
+            }
+            if (!toolTipString.equals(""))
+                toolTipString += ", ";
+            toolTipString += caseName;
+        }
+        NumberOnlyBuildLabel label = new NumberOnlyBuildLabel(testResultAction.run);
+        dataset.add(count, seriesName, label);
+
+        /*
+         * Also being stored in {@link #failToolTip/#skipToolTip/#totalToolTip} in order to generate tooltips on
+         * hovering mouse over the trend.
+         */
+        if (seriesName.equals("failed"))
+            failToolTip.put(label, toolTipString);
+        else if (seriesName.equals("skipped"))
+            skipToolTip.put(label, toolTipString);
+        else
+            totalToolTip.put(label, toolTipString);
+    }
+
+    /**
+     * A method to build dataset for the chosen project to generate trends.
+     * @param req The HTTP request message for the particular project or all projects for overall
+     *            build analysis trend type.
+     * @return An object of type {@link CategoryDataset} in which columns are the build numbers to be
+     * displayed on x-axis and rows are the different data series that need to be analysed for the chosen
+     * project.
+     * <p>
+     * This method creates {@link CategoryDataset} object for the chosen project with three data series
+     * namely "failed" for the no. of failed testcases, "skipped" for number of skipped testcases and
+     * "total" which contains no. of passed testcases. As the generated chart is stacked area chart and
+     * total is plotted on top of failed and skipped data series so, total(data series) effectively
+     * depict total number of testcases in the build.
+     */
+    private CategoryDataset buildDataSetPerProject (StaplerRequest req) {
+        boolean failureOnly = Boolean.valueOf(getParameter(req, FAILUREONLY));
+        String projectLevel = getParameter(req, PROJECTLEVEL);
+        boolean allPackages = projectLevel.equals(ALLPROJECTS);
+        DataSetBuilder<String, NumberOnlyBuildLabel> dataset = new DataSetBuilder<String, NumberOnlyBuildLabel>();
+        int cap = Integer.getInteger(AbstractTestResultAction.class.getName() + ".test.trend.max", Integer.MAX_VALUE);
+        int count = 0;
+        failToolTip = new ConcurrentHashMap<>();
+        skipToolTip = new ConcurrentHashMap<>();
+        totalToolTip = new ConcurrentHashMap<>();
```

Comment:
I think this won't be needed after you adress my previous change.

Reply:
Can't initialize them at the time of declaration as they are private variables of an abstract class and won't get initialized on creating an object of implementing class.